tripal/tripal
# Much improved GFF3 loader

# New Feature

Issue #1040

## Description


The GFF3 loader is unbearably slow when Chado begins to fill up.  It can take days to load a genome.  This fix greatly improves the performance and adds a quite a few functional tests.

## Testing?





To test the GFF3 loader you can use any GFF3 file available to you, but it is best if you have one that contains as many of the GFF3 acceptable attributes (Dbxref, Alias, Target, Note, Ontology_Term).  

1. Create a new organism (if it doesn't already exist) for the GFF3 data
2. Create a new analysis (if it doesn't already exist) for the GFF3 data
3. Use the GFF3 loader to import the GFF3 file.
4. After import of the GFF3 file, publish the features of interest (e.g. genes, mRNA)
5. Review the pages to ensure all relevant details of your GFF3 file are present.

The loader takes about 5 minutes to import a GFF3 file with 1 million features.

You should see output similar to the following

```
2020-12-17 19:41:25
Tripal Job Launcher
Running as user 'admin'
-------------------
2020-12-17 19:41:25: There are 1 jobs queued.
2020-12-17 19:41:25: Job ID 214.
2020-12-17 19:41:25: Calling: tripal_run_importer(39)

Running 'Chado GFF3 File Loader' importer
NOTE: Loading of file is performed using a database transaction. 
If it fails or is terminated prematurely then all insertions and 
updates are rolled back and will not be found in the database

Opening /var/www/html/sites/default/files/tripal/users/1/Hebr.1_0.gff
Opening temporary cache file: /tmp/TripalGFF3Import_zXc0QF
Step  1 of 26: Caching GFF3 file...                             
Step  2 of 26: Find existing landmarks...                         
Step  3 of 26: Insert new landmarks (if needed)...                
Step  4 of 26: Find missing proteins...                           
Step  5 of 26: Add missing proteins to list of features...        
Step  6 of 26: Find existing features...                          
Step  7 of 26: Clear attributes of existing features...            
Step  8 of 26: Processing 1,020,864 features...               
Step  9 of 26: Get new feature IDs...                             
Step 10 of 26: Insert locations...                               
Step 11 of 26: Associate parents and children...                 
Step 12 of 26: Calculate child ranks...                          
Step 13 of 26: Add child-parent relationships...                 
Step 14 of 26: Insert properties...                               
Step 15 of 26: Find synonyms (aliases)...                         
Step 16 of 26: Insert new synonyms (aliases)...                  
Step 17 of 26: Insert feature synonyms (aliases)...              
Step 18 of 26: Find cross references...                          
Step 19 of 26: Insert new cross references...                    
Step 20 of 26: Get new cross references IDs...                   
Step 21 of 26: Insert feature cross references...                
Step 22 of 26: Insert feature ontology terms...                  
Step 23 of 26: Insert 'derives_from' relationships...            
Step 24 of 26: Insert Targets...                                 
Step 25 of 26: Associate features with analysis....              
Step 26 of 26: Adding sequences data (Skipped: none available)...

Done.
Committing Transaction...
```

## Screenshots (if appropriate):

## Additional Notes (if any):
